wp-includes/mysql/class-mysql-lexer.php

```diff
@@ -0,0 +1,2752 @@
+				$this->type     = self::IDENTIFIER;
+			}
+		}
+	}
+
+	protected function single_quoted_text() {
+		$pattern = $this->is_sql_mode_active( self::SQL_MODE_NO_BACKSLASH_ESCAPES )
+			? self::PATTERN_SINGLE_QUOTED_TEXT_NO_BACKSLASH_ESCAPES
+			: self::PATTERN_SINGLE_QUOTED_TEXT;
+
+		if ( preg_match( '/\G' . $pattern . '/u', $this->input, $matches, 0, $this->position ) ) {
+			$this->text      = $matches[0];
+			$this->position += strlen( $this->text );
+			$this->c         = $this->input[ $this->position ] ?? null;
+			$this->n         = $this->input[ $this->position + 1 ] ?? null;
+			$this->type      = self::SINGLE_QUOTED_TEXT;
+		} else {
+			$this->consume();
+			$this->type = self::INVALID_INPUT;
+		}
+	}
+
+	protected function double_quoted_text() {
+		$pattern = $this->is_sql_mode_active( self::SQL_MODE_NO_BACKSLASH_ESCAPES )
+			? self::PATTERN_DOUBLE_QUOTED_TEXT_NO_BACKSLASH_ESCAPES
+			: self::PATTERN_DOUBLE_QUOTED_TEXT;
+
+		if ( preg_match( '/\G' . $pattern . '/u', $this->input, $matches, 0, $this->position ) ) {
+			$this->text      = $matches[0];
+			$this->position += strlen( $this->text );
+			$this->c         = $this->input[ $this->position ] ?? null;
+			$this->n         = $this->input[ $this->position + 1 ] ?? null;
+			$this->type      = self::DOUBLE_QUOTED_TEXT;
+		} else {
+			$this->consume();
+			$this->type = self::INVALID_INPUT;
+		}
+	}
+
+	protected function back_tick_quoted_id() {
+		$pattern = $this->is_sql_mode_active( self::SQL_MODE_NO_BACKSLASH_ESCAPES )
+			? self::PATTERN_BACKTICK_QUOTED_ID_NO_BACKSLASH_ESCAPES
+			: self::PATTERN_BACKTICK_QUOTED_ID;
+
+		if ( preg_match( '/\G' . $pattern . '/u', $this->input, $matches, 0, $this->position ) ) {
+			$this->text      = $matches[0];
+			$this->position += strlen( $this->text );
+			$this->c         = $this->input[ $this->position ] ?? null;
+			$this->n         = $this->input[ $this->position + 1 ] ?? null;
+			$this->type      = self::BACK_TICK_QUOTED_ID;
+		} else {
+			$this->consume();
+			$this->type = self::INVALID_INPUT;
+		}
+	}
+
+	protected function hex_number() {
+		$is_quoted = 'x' === strtolower( $this->c ) && "'" === $this->n;
+
+		// Consume "0x" or "x'".
+		$this->consume();
+		$this->consume();
+
+		while (
+			( $this->c >= '0' && $this->c <= '9' )
+			|| ( $this->c >= 'a' && $this->c <= 'f' )
+			|| ( $this->c >= 'A' && $this->c <= 'F' )
+		) {
+			$this->consume();
+		}
+
+		if ( $is_quoted ) {
+			$this->consume(); // Consume the "'".
+		}
+
+		$this->type = self::HEX_NUMBER;
+	}
+
+	protected function bin_number() {
+		$is_quoted = 'b' === strtolower( $this->c ) && "'" === $this->n;
+
+		// Consume "0b" or "b'".
+		$this->consume();
+		$this->consume();
+
+		while ( '0' === $this->c || '1' === $this->c ) {
+			$this->consume();
+		}
+
+		if ( $is_quoted ) {
+			$this->consume(); // Consume the "'".
+		}
+
+		$this->type = self::BIN_NUMBER;
+	}
+
+	protected function int_number() {
+		while ( $this->is_digit( $this->c ) ) {
```

Comment:
`strspn()` would be faster, we could introduce a helper method to reconcile the rest of the state after a `strspn()` call.